### ISSUE TITLE
shmem input crashes with segfault when connecting to squeezelite

### ISSUE THREAD

**Issue by chincheta0815** [karlstav/cava#418]
Using the latest source from the github.com master branch, cava crashes on MacOS Monterey when connecting to a local squeezelite instance via shmem. Using cava without shmem / setting it to use portaudio works fine.

Seemingly random, I get
 `input_shmem: source: /squeezelite-ac:bc:32:7a:2d:99cava(20773,0x1080b0600) malloc: Incorrect checksum for freed object 0x7fef8303dc00: probably modified after being freed.
Corrupt value: 0x4061a00000000000
cava(20773,0x1080b0600) malloc: *** set a breakpoint in malloc_error_break to debug
zsh: abort      cava
`

I hope I can give the correct information to reproduce below.
Otherwise please give me hint concerning what I am doing wrong...

**To Reproduce**
0. Start squeezelite with vis output
1. Configure cava to use shmem with noncurses
2. run cava
3. wait

**Expected behavior**
No error ;o)

**Desktop (please complete the following information):**
 - OS: macOS
 - Version 15.1

**Terminal emulator**
- MacOS Terminal with zsh

**Additional context**
The latest output_vis.c of squeezelite is here (https://github.com/ralph-irving/squeezelite/blob/master/output_vis.c). Maybe that helps.

**Comment by chincheta0815** [comment]
I found some interesting lines that might have an influence here: https://github.com/ralph-irving/jivelite/blob/master/src/visualizer/vumeter.c

Is there the need to rdlock the vis_map?

I am not too deep into programming and thread frighten me.
Using two lock when filling the fftw buffers seems very special and might cause deadlocks...

**Comment by karlstav** [comment]
not sure how to debug, don't realy have any squeezelite stuff. and definently not having anything that can run macos 15.

maybe check with @atisharma who was the last to fix the shmem input.

**Comment by chincheta0815** [comment]
@atisharma introduced some more changes to shmem input. that might be an idea. in the file im mentioned i found some stuff that rdlocks the mmap...
so far it looks that there is some buffer empty or so...
but i lack knowledge in debugging threaded code...

**Comment by atisharma** [comment]
The only thing I can contribute is my ignorance - I have very limited knowledge about shared memory implementations and no knowledge of or access to OS X, sorry.

**Comment by karlstav** [comment]
hi @chincheta0815.

sorry about the delay here. There was indeed a very ugly buffer overflow in the shmem input module, that I introduced a while back. I have pushed a fix for it, can you test it?

I have tested that it no longer crashes. But don't know how to actually test it with audio coming through. Do i need some special hardware, or can i test it with just squeezelight and this Logitech Media Server software?

**Comment by docgalaxyblock** [comment]
Hi @karlstav 
I'm not sure if I need to compile it from source or the debian repo already has the newest version...
`cava 0.7.3-1-g80d465f`
If I run cava with `squeezelite -v` I only get a `Segmentation fault`
I might have time this evening to compile it from source if this version is not up to date

Yes you can test it with LMS + squeezelite and listen to something 😉
Thanks

**Comment by karlstav** [comment]
Yes you will need to get the latest source from here and compile it.

The changes will make it into the debian repo. But it can take some time.

I will try to set up LMS and squeezelight. So I can test stuff before braking it next time.

**Comment by docgalaxyblock** [comment]
Ok I compile it if I have time.

Do you want to use a vm or a pi for testing?
If a pi maybe use DietPi it has a sofware manager where you can edit the dietpi.txt after flashing the imgage to setup auto install.
LMS would be ID35 , squeezelite ID36, build essentials ID16, git ID17

And its squeeze`lite` not squeeze`light` 😜

**Comment by atisharma** [comment]
I just compiled and ran the latest commit, on Void linux x86_64.
It runs with squeezelite shmem without crashing. But, the display looks wrong. I compare with pulseaudio which looks correct.

Karl - to save you a bit of time, if you want to compile squeezelite you'll need to do it with
`OPTS="-DVISEXPORT" make --makefile=Makefile.pulse`
(it took me a bit of hunting to find the right options).

**Comment by karlstav** [comment]
looks like there is a package for it in ubuntu 21.10 with the visexport stuff enabled. will check it out.

@Jocker666z already pointed out the noise issue in #375, let's move the discussion over here.

I am looking it to it now

**Comment by karlstav** [comment]
try it now, I was a bit quick. The buffer must of course be frames * 2 (left and right).

**Comment by Jocker666z** [comment]
> try it now, I was a bit quick. The buffer must of course be frames * 2 (left and right).

https://github.com/karlstav/cava/commit/1b055fa056b7cc44bb85b8743a691f6ce58e4228 fixed noise, perfect :)

**Comment by atisharma** [comment]
Works for me too :)

**Comment by karlstav** [comment]
but not for me :(

I got the LMS and squeezelite up and running. music is playing, but the visualizer is only refreshing like once every 2 seconds. hmm. is there some buffer settings somewhere?

**Comment by atisharma** [comment]
I don't think so... could be a bug in an old version of squeezelite?
Does the audio play OK?

**Comment by karlstav** [comment]
Squeezelite v1.9.8-1317, audio plays fine. 

I am using squeezelite-pulseaudio.

**Comment by Jocker666z** [comment]
Squeezelite v1.9.8-1317 with alsa here

**Comment by karlstav** [comment]
hm, same problem with alsa

**Comment by atisharma** [comment]
Try with a different music file perhaps?

**Comment by karlstav** [comment]
same with different kinds and also radio stations.

what are your Build options? bottom of `squeezelite -v`output

```
Build options: LINUX PULSEAUDIO EVENTFD RESAMPLE FFMPEG VISEXPORT IR DSD SSL LINKALL
```

**Comment by Jocker666z** [comment]
> I just compiled and ran the latest commit, on Void linux x86_64. It runs with squeezelite shmem without crashing. But, the display looks wrong. I compare with pulseaudio which looks correct.
> 
> Karl - to save you a bit of time, if you want to compile squeezelite you'll need to do it with `OPTS="-DVISEXPORT" make --makefile=Makefile.pulse` (it took me a bit of hunting to find the right options).

**Comment by atisharma** [comment]
Build options: LINUX PULSEAUDIO EVENTFD VISEXPORT DSD

**Comment by chincheta0815** [comment]
Hej there,

you were too fast ;o)
Once everything works please send a hint to me.
Then I could arrange an update for the homebrew formular.

**Comment by karlstav** [comment]
@chincheta0815, it is working fine. I was just struggling with making squeezelite visualizer output working for me. But it doesn't matter. I will draft a new release soon (hopefully later today). Then you can use that in the formular.

**Comment by karlstav** [comment]
@chincheta0815, released new version 0.7.5
